src/TLuaInterpreter.cpp

```diff
@@ -10739,6 +10750,505 @@ int TLuaInterpreter::restartIrc(lua_State* L)
+    lua_pushstring(L, currentVoice.toUtf8().constData());
+    return 1;
+}
+
+
+/** ttsSetVoiceByName( name )
+ *  Sets the current speechUnit voice if given a name.
+ *  Returns true if set, false otherwise.
+ */
+int TLuaInterpreter::ttsSetVoiceByName(lua_State* L)
+{
+    TLuaInterpreter::ttsBuild();
+    
+    QString nextVoice;
+    
+    if (!lua_isstring(L, 1)) {
+        lua_pushfstring(L, "ttsSetVoiceByName: argument expected string, got %s!", luaL_typename(L, 1));
+        lua_error(L);
+        return 1;
+    } else {
+        nextVoice = QString(lua_tostring(L, 1));
+    }
+    
+    QVector<QVoice> speechVoices = speechUnit->availableVoices();
+    foreach (const QVoice &voice, speechVoices) {
+        if(voice.name() == nextVoice)
+        {
+            speechUnit->setVoice(voice);
+            lua_pushboolean(L, true);
+            return 1;
+        }
+    }
+    
+    lua_pushboolean(L, false);
+    return 1;
+}
+
+
+/** ttsSetVoiceByIndex( index )
+ *  Sets the current speechUnit voice if given a proper index.
+ *  Returns true if set, errors otherwise.
+ *  Should only be used in conjunction with ttsGetVoices.
+ */
+int TLuaInterpreter::ttsSetVoiceByIndex(lua_State* L)
+{
+    TLuaInterpreter::ttsBuild();
+    
+    int index;
+    if (!lua_isnumber(L, 1)) {
+        lua_pushfstring(L, "ttsSetVoiceByIndex: argument expected number, got %s!", luaL_typename(L, 1));
+        lua_error(L);
+        return 1;
+    } else {
+        index = lua_tonumber(L, 1);
+    }
+    
+    index--;
+    
+    QVector<QVoice> speechVoices = speechUnit->availableVoices();
+    if (index < 0 || index > speechVoices.size())
+    {
+        lua_pushstring(L, "ttsSetVoiceByIndex: voice index out of bounds");
+        lua_error(L);
+        return 1;
+    }
+    
+    speechUnit->setVoice(speechVoices.at(index));
+    
+    return 0;
+}
+
+/**
+ *  Handles playback of queued text.
+ */
+void TLuaInterpreter::ttsStateChanged(QTextToSpeech::State state)
+{
+    TEvent event;
+    switch(state)
+    {
+        case QTextToSpeech::Paused : event.mArgumentList.append(QLatin1String("ttsSpeechPaused")); break;
+        case QTextToSpeech::Speaking : event.mArgumentList.append(QLatin1String("ttsSpeechStarted")); break;
+        case QTextToSpeech::BackendError : event.mArgumentList.append(QLatin1String("ttsSpeechError")); break;
+        case QTextToSpeech::Ready : event.mArgumentList.append(QLatin1String("ttsSpeechReady")); break;
+    }
+    event.mArgumentTypeList.append(ARGUMENT_TYPE_STRING);
+    
+    if (state == QTextToSpeech::Speaking)
+    {
+        event.mArgumentList.append(speechCurrent);
+        event.mArgumentTypeList.append(ARGUMENT_TYPE_STRING);
+    }
+    
+    mudlet::self()->getHostManager().postInterHostEvent(NULL, event, true);
+    
+    if(state != QTextToSpeech::Ready || speechQueue.empty())
+    {
+        return;
+    }
+    
+    speechCurrent = speechQueue.takeFirst();
+    
+    speechUnit->say(speechCurrent);
```

Comment:
I can't really express it, but this looks like a code smell. This function should only raise the event, but you want more control over when you really want to start the playback (probably only when something was queued and noting is played yet).